server/rosetta/cosmos/client/client.go

```diff
@@ -0,0 +1,562 @@
+
+// WithCodec allows to build a client with a predefined interface registry and codec
+func WithCodec(ir codectypes.InterfaceRegistry, cdc *codec.ProtoCodec) OptionFunc {
+	return func(o *options) {
+		o.interfaceRegistry = ir
+		o.cdc = cdc
+	}
+}
+
+// Client implements a single network client to interact with cosmos based chains
+type Client struct {
+	auth auth.QueryClient
+	bank bank.QueryClient
+
+	ir codectypes.InterfaceRegistry
+
+	clientCtx client.Context
+}
+
+func (c *Client) SupportedOperations() []string {
+	var supportedOperations []string
+	for _, ii := range c.ir.ListImplementations("cosmos.base.v1beta1.Msg") {
+		resolve, err := c.ir.Resolve(ii)
+		if err == nil {
+			if _, ok := resolve.(rosetta.Msg); ok {
+				supportedOperations = append(supportedOperations, strings.TrimLeft(ii, "/"))
+			}
+		}
+	}
+
+	supportedOperations = append(supportedOperations, rosetta.OperationFee)
+
+	return supportedOperations
+}
+
+func (c *Client) PreprocessOperationsToOptions(ctx context.Context, req *types.ConstructionPreprocessRequest) (options map[string]interface{}, err error) {
+	operations := req.Operations
+	if len(operations) > 3 {
+		return nil, rosetta.ErrInvalidRequest
+	}
+
+	msgs, err := conversion.ConvertOpsToMsgs(c.ir, operations)
+	if err != nil {
+		return nil, rosetta.WrapError(rosetta.ErrInvalidAddress, err.Error())
+	}
+
+	if len(msgs) < 1 || len(msgs[0].GetSigners()) < 1 {
+		return nil, rosetta.WrapError(rosetta.ErrInterpreting, "invalid msgs from operations")
+	}
+
+	memo, ok := req.Metadata["memo"]
+	if !ok {
+		memo = ""
+	}
+
+	defaultGas := float64(200000)
+
+	gas := req.SuggestedFeeMultiplier
+	if gas == nil {
+		gas = &defaultGas
+	}
+
+	return map[string]interface{}{
+		rosetta.OptionAddress: msgs[0].GetSigners()[0],
+		rosetta.OptionMemo:    memo,
+		rosetta.OptionGas:     gas,
+	}, nil
+}
+
+func (c *Client) ConstructionPayload(ctx context.Context, request *types.ConstructionPayloadsRequest) (resp *types.ConstructionPayloadsResponse, err error) {
+	if len(request.Operations) > 3 {
+		return nil, rosetta.ErrInvalidOperation
+	}
+
+	msgs, fee, err := conversion.RosettaOperationsToSdkMsg(c.ir, request.Operations)
+	if err != nil {
+		return nil, rosetta.WrapError(rosetta.ErrInvalidOperation, err.Error())
+	}
+
+	metadata, err := services.GetMetadataFromPayloadReq(request)
+	if err != nil {
+		return nil, rosetta.WrapError(rosetta.ErrInvalidRequest, err.Error())
+	}
+
+	txFactory := tx.Factory{}.WithAccountNumber(metadata.AccountNumber).WithChainID(metadata.ChainID).
+		WithGas(metadata.Gas).WithSequence(metadata.Sequence).WithMemo(metadata.Memo).WithFees(fee.String())
+
+	TxConfig := c.getTxConfig()
+	txFactory = txFactory.WithTxConfig(TxConfig)
+
+	txBldr, err := tx.BuildUnsignedTx(txFactory, msgs...)
+	if err != nil {
+		return nil, err
+	}
+
+	if txFactory.SignMode() == signing.SignMode_SIGN_MODE_UNSPECIFIED {
+		txFactory = txFactory.WithSignMode(signing.SignMode_SIGN_MODE_LEGACY_AMINO_JSON)
```

Comment:
q: why legacy amino the default? this should also be declared on the godoc comment on top of the func

Reply:
its because the current sign mode direct needs online functionality that cannot be supported in an offline endpoint 

cc @sahith-narahari  for confirmation

Reply:
yes, currently sign_mode_direct needs the signers to be set before hand which isn't possible with rosetta flow of operations.

I'll create an issue for this and link it here.